lib/chef/provider/apt_repository.rb

```diff
@@ -0,0 +1,255 @@
+require "chef/provider/noop"
+
+class Chef
+  class Provider
+    class AptRepository < Chef::Provider
+      use_inline_resources
+
+      include Chef::Mixin::ShellOut
+
+      provides :apt_repository do
+        uses_apt?
+      end
+
+      def whyrun_supported?
+        true
+      end
+
+      def load_current_resource
+      end
+
+      action :add do
+        unless new_resource.key.nil?
+          if is_key_id?(new_resource.key) && !has_cookbook_file?(new_resource.key)
+            install_key_from_keyserver
+          else
+            install_key_from_uri
+          end
+        end
+
+        declare_resource(:execute, "apt-cache gencaches") do
+          ignore_failure true
+          action :nothing
+        end
+
+        declare_resource(:apt_update, new_resource.name) do
+          ignore_failure true
+          action :nothing
+        end
+
+        components = if is_ppa_url?(new_resource.uri) && new_resource.components.empty?
+                       "main"
+                     else
+                       new_resource.components
+                     end
+
+        repo = build_repo(
+          new_resource.uri,
+          new_resource.distribution,
+          components,
+          new_resource.trusted,
+          new_resource.arch,
+          new_resource.deb_src
+        )
+
+        declare_resource(:file, "/etc/apt/sources.list.d/#{new_resource.name}.list") do
+          owner "root"
+          group "root"
+          mode "0644"
+          content repo
+          sensitive new_resource.sensitive
+          action :create
+          notifies :run, "execute[apt-cache gencaches]", :immediately
+          notifies :update, "apt_update[#{new_resource.name}]", :immediately if new_resource.cache_rebuild
+        end
+      end
+
+      action :remove do
+        if ::File.exist?("/etc/apt/sources.list.d/#{new_resource.name}.list")
+          converge_by "Removing #{new_resource.name} repository from /etc/apt/sources.list.d/" do
+            declare_resource(:file, "/etc/apt/sources.list.d/#{new_resource.name}.list") do
+              sensitive new_resource.sensitive
+              action :delete
+              notifies :update, "apt_update[#{new_resource.name}]", :immediately if new_resource.cache_rebuild
+            end
+
+            declare_resource(:apt_update, new_resource.name) do
+              ignore_failure true
+              action :nothing
+            end
+
+          end
+        end
+      end
+
+      def self.uses_apt?
+        ENV["PATH"] ||= ""
+        paths = %w{ /bin /usr/bin /sbin /usr/sbin } + ENV["PATH"].split(::File::PATH_SEPARATOR)
+        paths.any? { |path| ::File.executable?(::File.join(path, "apt-get")) }
+      end
+
+      def is_key_id?(id)
+        id = id[2..-1] if id.start_with?("0x")
+        id =~ /^\h+$/ && [8, 16, 40].include?(id.length)
+      end
+
+      def extract_fingerprints_from_cmd(cmd)
+        so = shell_out(cmd, env: { "LANG" => "en_US", "LANGUAGE" => "en_US" })
```

Comment:
i think you can actually drop the lang stuff since this code should have you completely covered:

https://github.com/chef/chef/blob/master/chef-config/lib/chef-config/config.rb#L872-L922

including handling the case of servers that have no `en_US.utf-8` or `C.utf-8` support and falls back to C and yells at the user:

https://github.com/chef/chef/blob/master/chef-config/lib/chef-config/config.rb#L911